### ISSUE TITLE
Cannot select "Font Family" on Windows builds

### ISSUE THREAD

**Issue by FirasH** [synfig/synfig#1766]
**Synfig version & platform**:
1.5.0 Artifact build of 5 October 4699e374 (got from a pull request)

**Issue description**:
1. Create a new animation
2. Insert a `Text Layer` on Canvas
3. Change the `Font Family` value in the `Parameters Panel`

The dropdown menu is greyed out and it's not possible to select any fonts.
I could not use it the "old" way i.e. by copying the font file name in the `Font Family` box.

**Comment by DhairyaBahl** [comment]
@ice0 I would like to work on this issue. Kindly assign it to me.

Thanks :)

**Comment by DhairyaBahl** [comment]
@ice0 My build is giving some error like libatkmm dll missing !! what should I do ??

**Comment by ice0** [comment]
@DhairyaBahl you need to run it from MSYS2 console. Or do install step (which copies all required dll's to your install folder)

**Comment by DhairyaBahl** [comment]
@ice0 I was running it in msys2-mingw64.exe and I've installed all the dependencies using ./1-setup-windows.sh command. Do i have to repeat it ? and during running that it gives error 

(Stripping trailing CRs from patch; use --binary to disable.)
patching file /mingw64/include/libintl.h
Reversed (or previously applied) patch detected!  Assume -R? [n]

Can this be a case ??

Thanks :)

**Comment by ice0** [comment]
`./1-setup-windows.sh` need to be called only once. After all dependencies was installed no need to run it.
If you run synfigstudio from MSYS shell (better to use some file manager like Midnight commander), then please check your `/mingw64/bin/` folder. `libatk-1.0-0.dll` should be here.

But. And this is important. If you trying to run synfigstudio from Windows explorer then you need to copy all dependencies to Synfig's `bin` folder. This is not recommended yet.

**Comment by DhairyaBahl** [comment]
@ice0 Ohh yes, I totally forgot. Last time I executed it in MSYS2 shell. This time I was trying directly using windows explorer!!

Thanks :) its working now :)

**Comment by DhairyaBahl** [comment]
@ice0 @Keyikedalube I was exploring synfig/synfig-studio/src/gui/widgets/widget_fontfamily.h and fontfamily.cpp and i am studying about gtkmm combo-boxes !! Kindly tell me if I am going in right direction ??

Thanks :)

**Comment by ice0** [comment]
I think yes.

**Comment by DhairyaBahl** [comment]
@Keyikedalube @ice0 @FirasH Can you guys give me a hint or something. I am stuck, I understand gtkmm combo boxes but not in detail that I can debug this issue. Kindly guide me a bit. 

Thank you :)

**Comment by ice0** [comment]
If I understand correctly, the version built with `2-build-debug.sh` script worked fine. If this is true, then I recommend that you attach to the executable using a debugger (gdb) and walk through the code step by step to understand the difference between them.

**Comment by DhairyaBahl** [comment]
@ice0 Thanks for the help. I will now use the debugger as well. 

Thank you :)

**Comment by rodolforg** [comment]
There are two different problems listed in this issue, both on Windows:
1. The combobox for Font `family` parameter of `Text` layer doesn't list font families;
2. Font `family` parameter of `Text` layer doesn't support font filenames.

The combobox is a custom widget: Widget_FontFamily, and it is filled only if OS building provides support for `fontconfig` library.
Code is surrounded with a guard for it. A first step would be investigate if it builds with fontconfig support (maybe adding a temporary [#error or #warning directive](https://gcc.gnu.org/onlinedocs/cpp/Diagnostics.html) to check it on building time).
Code is here:
https://github.com/synfig/synfig/blob/master/synfig-studio/src/gui/widgets/widget_fontfamily.cpp

If parameter is not accepting the filename input, I see two possible reasons:

1. Somehow I messed up the text layer when fixing other stuff – including fontconfig support (https://github.com/synfig/synfig/blob/master/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp);
2. The combobox text is not being 'sent' to synfig layer, it isn't really changing the font family parameter of Text layer. Maybe events are not properly signalized or connected.

**Comment by DhairyaBahl** [comment]
![IMG_20201018_201238.jpg](https://user-images.githubusercontent.com/71627983/96371802-4326c600-1181-11eb-9964-50fc63086aeb.jpg)

@rodolforg @ice0 

It is able to get FontConfig support as I can see in this build log which I copied from cmder. 

Thanks for helping

**Comment by DhairyaBahl** [comment]
@rodolforg @FirasH @ice0 

Work Update 
I just tried entering names of various fonts myself just to be sure about the working of font family parameter of the text layer and its working . We just have to find why that dropbox is not working and listing the fonts. 

These are the fonts I tried : Helvetica, ariel, Garamond, serif, sans serif( default one ) and calibri. 

Thanks for help :)

**Comment by rodolforg** [comment]
> The combobox is a custom widget: Widget_FontFamily, and it is filled only if OS building provides support for `fontconfig` library.
> Code is surrounded with a guard for it. A first step would be investigate if it builds with fontconfig support (maybe adding a temporary [#error or #warning directive](https://gcc.gnu.org/onlinedocs/cpp/Diagnostics.html) to check it on building time).
> Code is here:
> https://github.com/synfig/synfig/blob/master/synfig-studio/src/gui/widgets/widget_fontfamily.cpp

Please check this file

**Comment by DhairyaBahl** [comment]
@rodolforg @ice0 @FirasH 

I added #warning directive in the beginning of the function refresh() and I expected that warning to appear on the screen during its build but I didn't get any warnings that mean function refresh() is not being called properly. Now I think its not getting proper font config support as it is not able to call that function !!  Kindly tell me if I am going in the right direction. 

Thanks for the help :)

**Comment by ice0** [comment]
Yes! You are in the right direction. Now you need to understand why `WITH_FONTCONFIG` is not set properly (as it should be). Hint: take a look at the CMakeLists.txt files.

**Comment by DhairyaBahl** [comment]
@ice0 The only fishy thing I found in the CMakeLists.ext files is missing "REQUIRED" in the font-config parameter and I added it in the pkg_check_modules . Is there anything else to fix or should I start a test-build now ? 

I've been doing analysis of all the cmakelists.txt files but there is nothing else except this which looked fishy to me but I don't think it will rectify the problem. Kindly tell me if there is anything else to rectify.  I am asking because it takes more than 2 hours for my pc to build synfig again :sweat_smile: .

Thanks :+1:

**Comment by ice0** [comment]
If you use "MSYS Makefiles" generator you can use `make --dry-run` flag to skip actual building.
For Ninja generator this will be `ninja -n --verbose`.

This is what I got using next commands:
```
mkdir test-build1 && cd test-build1
cmake ../synfig -GNinja
ninja -n --verbose | grep -i widget_font
```

All takes near 1 minute maximum.

And this is the output (I made a small formatting here):
**ninja -n --verbose | grep -i widget_font**
```
[884/906] C:\msys64\mingw64\bin\c++.exe  
-DCMAKE_BUILD 
-DDEVEL_VERSION="\"Revision date: 2020-10-21\nBranch: master\nRevision ID: fa4021de0\"" 
-DETL_VERSION=\"1.5.0\" 
-DHAVE_CONFIG_H 
-DIMAGE_DIR="\"C:/Program Files (x86)/synfig-studio/share/pixmaps\"" 
-DIMAGE_EXT=\"png\" 
-DLOCALEDIR="\"C:/Program Files (x86)/synfig-studio/share/locale\"" 
-DSHOW_EXTRA_INFO 
-DSYNFIG_DATADIR="\"C:/Program Files (x86)/synfig-studio/share/synfig\"" 
-IC:/msys64/home/u/synfig/ETL -IETL 
-IC:/msys64/home/u/synfig/synfig-core/src -Isynfig-studio/src 
-IC:/msys64/home/u/synfig/synfig-studio/src -Isynfig-studio/src/gui 
-IC:/msys64/home/u/synfig/synfig-studio/src/gui -isystem C:/msys64/mingw64/include/sigc++-2.0 -isystem C:/msys64/mingw64/lib/sigc++-2.0/include 
-isystem C:/msys64/mingw64/include/gtkmm-3.0 -isystem C:/msys64/mingw64/lib/gtkmm-3.0/include -isystem C:/msys64/mingw64/include/atkmm-1.6 
-isystem C:/msys64/mingw64/include/gdkmm-3.0 
...
-g -fno-inline-functions -fno-inline -g -O0 -Wall   -std=gnu++11 -MD -MT 
synfig-studio/src/gui/CMakeFiles/synfigstudio.dir/widgets/widget_fontfamily.cpp.obj 
-MF synfig-studio\src\gui\CMakeFiles\synfigstudio.dir\widgets\widget_fontfamily.cpp.obj.d 
-o synfig-studio/src/gui/CMakeFiles/synfigstudio.dir/widgets/widget_fontfamily.cpp.obj 
-c C:/msys64/home/u/synfig/synfig-studio/src/gui/widgets/widget_fontfamily.cpp
```

Did you see something suspicious here?

**Comment by ice0** [comment]
For example this is output for lyr_freetype

```
ninja -n --verbose | grep -i lyr_freetype
[74/906] C:\msys64\mingw64\bin\c++.exe  
-DHAVE_CONFIG_H 
-DWITH_FONTCONFIG 
-Dlyr_freetype_EXPORTS 
-IC:/msys64/home/u/synfig/ETL 
-IETL -IC:/msys64/mingw64/include/OpenEXR -IC:/msys64/mingw64/include/ImageMagick-7 -Isynfig-core/src -IC:/msys64/home/u/synfig/synfig-core/src 
...
-g -fno-inline-functions -fno-inline -g -O0 -Wall   -std=gnu++11 -MD 
-MT synfig-core/src/modules/lyr_freetype/CMakeFiles/lyr_freetype.dir/main.cpp.obj 
-MF synfig-core\src\modules\lyr_freetype\CMakeFiles\lyr_freetype.dir\main.cpp.obj.d 
-o synfig-core/src/modules/lyr_freetype/CMakeFiles/lyr_freetype.dir/main.cpp.obj 
-c C:/msys64/home/u/synfig/synfig-core/src/modules/lyr_freetype/main.cpp
```

**Comment by DhairyaBahl** [comment]
@ice0 

I tried to do test-build on my local machine using your given commands but I got some errors.
-- Configuring done
-- Generating done
-- Build files have been written to: C:/synfig/test-build1
make[2]: *** No rule to make target 'output/lib/libsynfig.dll.a', needed by 'output/lib/synfig/modules/liblyr_freetype.dll'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:1051: synfig-core/src/modules/lyr_freetype/CMakeFiles/lyr_freetype.dir/all] Error 2
make: *** [Makefile:172: all] Error 2

And the only error or flaw I found in your given output is DWITH_FONTCONFIG which is present in the lyr_freetype but not in the  widget_font. 

Thanks for help :)

**Comment by ice0** [comment]
> make: *** [Makefile:172: all] Error 2

Just use Ninja in a different folder as I wrote earlier.

> And the only error or flaw I found in your given output is DWITH_FONTCONFIG which is present in the lyr_freetype but not in the widget_font.

Yes, this is the reason, why widget not working.

**Comment by DhairyaBahl** [comment]
@ice0 Okay Thanks Its working now. Now I understand what you meant by u did some formatting. I should now work on why DWITH_FONTCONFIG is not present in widget_font also.

Thanks again :)

**Comment by DhairyaBahl** [comment]
@ice0 I compared the cmakelists files of both of them i.e. widget_fontfamily and lyr_freetype and I found that target_link libraries in the CMakeLists.txt of widget_fontfamily is not having support of ${FONT_CONFIG_LIBRARIES} . 

So if I am correct, this should solve the issue and will add the missing library. If yes, I will now run the build to test that. 

Thanks :)

**Comment by rodolforg** [comment]
> And the only error or flaw I found in your given output is DWITH_FONTCONFIG which is present in the lyr_freetype but not in the widget_font.

If that enables on Windows CMake building, we should check if it's working with autotools too.

Anyway, there would still be the second issue reported by @Firash in original post:

>  I could not use it the "old" way i.e. by copying the font file name in the `Font Family` box.

that would be a lyr_freetype issue I could dig later.

**Comment by DhairyaBahl** [comment]
@rodolforg I was able to change the font family by copying and pasting names of the font. If you want I can share the screenshots !!

**Comment by ice0** [comment]
@DhairyaBahl the real problem the WITH_FONTCONFIG flag. As you can notice, it sets value in `synfig\synfig-core\src\CMakeLists.txt`
```
if(FONT_CONFIG_FOUND)
  add_definitions(-DWITH_FONTCONFIG)
endif()
```

after that it out of scope (when returns to parent folder), and did not work when it begins to build `synfig\synfig-studio` files.

**Comment by DhairyaBahl** [comment]
@ice0 
1. why it is going out of scope when we are passing this "add_subdirectory(src)"  in the cmakelist of parent folder ?

2. 
if(FONT_CONFIG_FOUND)
add_definitions(-DWITH_FONTCONFIG)
endif() 
This is missing in the cmake list of synfig studio/src/ 

Kindly guide further. Thanks :)

**Comment by ice0** [comment]
> 1. why it is going out of scope when we are passing this "add_subdirectory(src)" in the cmakelist of parent folder ?

Because `add_definitions` works for current `CMakeLists.txt` folder and to child folders

> 2.

Yes

> 3.

Yes :)

**Comment by DhairyaBahl** [comment]
@ice0

I am able to get -DWITH_FONTCONFIG SUPPORT NOW !! YaY!! as you can see below. I've removed the unnecessary part of -isystem to make it look cleaner and shorter. Going to begin with test-build :) 

[805/906] C:\msys64\mingw64\bin\c++.exe  
-DCMAKE_BUILD -DDEVEL_VERSION="\"Revision date: 2020-10-21\nBranch: master\nRevision ID: fa4021de0\"" 
-DETL_VERSION=\"1.5.0\" 
-DHAVE_CONFIG_H 
-DIMAGE_DIR="\"C:/Program Files (x86)/synfig-studio/share/pixmaps\"" 
-DIMAGE_EXT=\"png\" 
-DLOCALEDIR="\"C:/Program Files (x86)/synfig-studio/share/locale\"" 
-DSHOW_EXTRA_INFO -DSYNFIG_DATADIR="\"C:/Program Files (x86)/synfig-studio/share/synfig\"" 
-DWITH_FONTCONFIG -IC:/synfig/ETL 

 -g -fno-inline-functions -fno-inline -g -O0 -Wall   -std=gnu++11 -MD -MT synfig-studio/src/gui/CMakeFiles/synfigstudio.dir/widgets/widget_fontfamily.cpp.obj -MF synfig-studio\src\gui\CMakeFiles\synfigstudio.dir\widgets\widget_fontfamily.cpp.obj.d -o synfig-studio/src/gui/CMakeFiles/synfigstudio.dir/widgets/widget_fontfamily.cpp.obj -c C:/synfig/synfig-studio/src/gui/widgets/widget_fontfamily.cpp

Thanks :)

**Comment by ice0** [comment]
@FirasH please check :)

**Comment by FirasH** [comment]
@ice0 the selector is now working!
The second method (name of the font file) still not working, but I assume that's a separate issue.

**Comment by DhairyaBahl** [comment]
@ice0 @FirasH 

I would like to tell that The second method of copy pasting the font name is working in my build..

I can share the screenshots.. of different fonts. 

Thanks :)

**Comment by FirasH** [comment]
@DhairyaBahl are you using Windows?
I get that issue only here.

**Comment by DhairyaBahl** [comment]
@FirasH yes, I am using windows build..

**Comment by DhairyaBahl** [comment]
![IMG_20201024_161834.jpg](https://user-images.githubusercontent.com/71627983/97079846-a3e84f80-1614-11eb-98d6-e04268bbf9a9.jpg)![IMG_20201024_161758.jpg](https://user-images.githubusercontent.com/71627983/97079849-a64aa980-1614-11eb-8b44-2bf0b9074686.jpg)![IMG_20201024_161736.jpg](https://user-images.githubusercontent.com/71627983/97079852-a8146d00-1614-11eb-806a-e85faebb34f5.jpg)

These are examples of 3 different fonts.Kindly look into these @FirasH  @ice0

**Comment by FirasH** [comment]
Yes, that is working because (I assume) the string matches the font listed in the list.

What I'm referring to is explicitly indicating the font file like `./font.otf` or `./font.ttf`

**Comment by DhairyaBahl** [comment]
@FirasH Ohh..Sorry IDK about that.

**Comment by ice0** [comment]
@rodolforg Can we use [GtkFontChooserWidget](https://developer.gnome.org/gtk3/stable/GtkFontChooserWidget.html) here?

**Comment by rodolforg** [comment]
@ice0, it doesn't allow to choose a font by filename. Some users use a specific font file in the animation folder, for example.

**Comment by rodolforg** [comment]
@DhairyaBahl thank you for your help!

I was able to track e squash the other bug of OP.
Can you check it now, @Firash?

**Comment by FirasH** [comment]
> 
> 
> @DhairyaBahl thank you for your help!
> 
> I was able to track e squash the other bug of OP.
> Can you check it now, @FirasH?

Now font files are opened, but only when the full absolute path is specified.
It won't get relative paths like `./my-font.otf` in the project's folder.

I use this function with realtive paths, which is very useful in case of editing on different machine and operating systems.

**Comment by rodolforg** [comment]
@FirasH even with the file saved?

**Comment by rodolforg** [comment]
Ok, I can reproduce it now :P

**Comment by FirasH** [comment]
Yes.
Just loaded a project made with an older revision of 1.5 where text objects had a realtive path to font files w.r.t. the project file.

**Comment by rodolforg** [comment]
@FirasH  I believe now it's really solved :)

**Comment by ice0** [comment]
@rodolforg PR?

**Comment by rodolforg** [comment]
@ice0 lol

**Comment by ice0** [comment]
@FirasH can you check if problem with relative paths is fixed?

**Comment by FirasH** [comment]
> @FirasH can you check if problem with relative paths is fixed?

With the last pull request by @rodolforg (#1833) relative fonts work, either with `./font.otf` or `font.otf`
